extra/hsa/demo.py

```diff
@@ -0,0 +1,248 @@
+  check(hsa.hsa_agent_get_info(agent, hsa.HSA_AGENT_INFO_NAME, ctypes.byref(buf := ctypes.create_string_buffer(256))))
+  print("dev:", ctypes.string_at(buf).decode()) # gfx1100 is here!
+
+  check(hsa.hsa_agent_get_info(agent, hsa.HSA_AGENT_INFO_QUEUE_MAX_SIZE, ctypes.byref(queue_size := ctypes.c_uint32())))
+  print("max queue size:", queue_size.value)
+
+  UINT32_MAX = (2 << 32) - 1
+  hsa_queue_ptr_t = ctypes.POINTER(hsa.hsa_queue_t)
+  null_func = ctypes.CFUNCTYPE(None, hsa.hsa_status_t, ctypes.POINTER(hsa.struct_hsa_queue_s), ctypes.POINTER(None))()
+  check(hsa.hsa_queue_create(agent, queue_size, hsa.HSA_QUEUE_TYPE_SINGLE, null_func, None, UINT32_MAX, UINT32_MAX, ctypes.byref(queue := hsa_queue_ptr_t())))
+
+  print("***** create HSA kernel")
+  a = Tensor([1.,2.]*128, device="HIP").realize()
+  b = Tensor([6.,5.]*128, device="HIP").realize()
+  binary, name = compile_ast_to_hip((a + b).realize()) # elf
+  kern = Kernel(agent, binary, name)
+
+  print("***** create simple args")
+  # char* in=(char*)malloc(1024*1024*4)
+  # memset(in, 1, 1024*1024*4)
+  # err=hsa_memory_register(in, 1024*1024*4)
+
+  # def simple_memory(sz, val):
+  #   mv = memoryview(bytearray(sz))
+  #   ctypes.memset(from_mv(mv), 0, len(mv))
+  #   check(hsa.hsa_memory_register())
+  #   return 
+
+
+  print("***** prep kernel")
+  check(hsa.hsa_signal_create(1, 0, None, ctypes.byref(signal := hsa.hsa_signal_t())))
+  
+  
+  kernarg_region = hsa.hsa_region_t()
+  kernarg_region.handle = -1
+  hsa.hsa_agent_iterate_regions(agent, filter_shared_memtype, ctypes.byref(kernarg_region))
+  print(kernarg_region.handle)
+
+  hip_dev = Device["HIP"]
+  hip_buffer = hip_dev.allocator.alloc(2 * 32 * 4 * 4)
+  hip_dev.synchronize()
+
+  class Args(ctypes.Structure): # This is what we've put in extra[1], seems that...
+    _fields_ = [("l1", ctypes.c_void_p)]
+    _align_ = 16
+  args = Args()
+  args.l1 = hip_buffer
+
+  check(hsa.hsa_memory_allocate(kernarg_region, ctypes.c_size_t(kern.kernargs_segment_size.value), ctypes.byref(kernarg_address := ctypes.c_void_p())))
+  ctypes.memmove(kernarg_address, ctypes.addressof(args), ctypes.sizeof(args))
+  print(kernarg_address)
+
+  print("***** prepearing kernel")
+  index = hsa.hsa_queue_load_write_index_relaxed(queue)
+  print("q indx:", index)
+
+  base_address = ctypes.cast(queue.contents.base_address, ctypes.POINTER(hsa.hsa_kernel_dispatch_packet_t))
+  dispatch_packet_addr = ctypes.addressof(base_address[index & (queue.contents.size - 1)])
+  dispatch_packet_ptr = ctypes.pointer(base_address[index & (queue.contents.size - 1)])
+  print("dispatch_packet_addr", dispatch_packet_addr)
+
+  dispatch_packet_ptr.contents.setup |= 1 << hsa.HSA_KERNEL_DISPATCH_PACKET_SETUP_DIMENSIONS
+
+  dispatch_packet_ptr.contents.workgroup_size_x = 32
+  dispatch_packet_ptr.contents.workgroup_size_y = 1
+  dispatch_packet_ptr.contents.workgroup_size_z = 1
+  dispatch_packet_ptr.contents.grid_size_x = 2 * 32 # opencl-like
```

Comment:
Cool, wonder if it supports non divisible ones

Reply:
Yeah, works fine on the simple example I ran. Seems it just masks the last lanes of the workgroup.